lib/site_cache.rb

```diff
@@ -0,0 +1,213 @@
+# frozen_string_literal: true
+
+require "monitor"
+
+# This is an LRU cache in two dimensions. There are limits on:
+#
+# * The global size of the cache,
+# * The per-site size of the cache,
+#
+# In order to enforce these limits, we need be able to efficiently get the next
+# least recently used item globally and per site. To do this, there are two
+# structures:
+#
+# `@flattened' is a Hash from [site, key] -> value
+# `@nested' is a Hash from site -> Hash from key -> value
```

Comment:
Instead of duplicating the values across two hashes leading to increase memory consumption, I think it might be possible for `@flattened` to use a Set to store the `[site, key]` tuple? The addition, retrieval and deletion operations will still be O(1).

Reply:
The values don't get duplicated. Both structures contain references to the same keys and values, so I don't think that's any less memory intensive. Maybe there's value in having a single source of truth. A ruby set, internally, is a hash, so just putting `true` for each value in the flattened representation would be equivalent.

Reply:
> The values don't get duplicated. Both structures contain references to the same keys and values

Sorry this is my bad. I looked this up and you are right that hash will just reference the same "value" object in memory. 

> Maybe there's value in having a single source of truth.

Do you think it will simplify the implementation if there is only one source for the value in the cache? From my perspective, I think it can help to simplify the implementation and make the cache easier to understand.

Reply:
I did do this.